load_tests/benchmarks/engine.py

```diff
@@ -0,0 +1,151 @@
+import subprocess
+import threading
+from typing import Dict, List
+
+import docker
+from docker.models.containers import Container
+from loguru import logger
+
+from benchmarks.utils import kill
+
+
+class InferenceEngineRunner:
+    def __init__(self, model: str):
+        self.model = model
+
+    def run(self, parameters: list[tuple]):
+        NotImplementedError("This method should be implemented by the subclass")
+
+    def stop(self):
+        NotImplementedError("This method should be implemented by the subclass")
+
+
+class TGIRunner(InferenceEngineRunner):
+    def __init__(self, model: str):
+        super().__init__(model)
+        self.process = None
+        self.model = model
+
+    def run(self, parameters: list[tuple]):
+        params = ""
+        for p in parameters:
+            params += f"--{p[0]} {str(p[1])}"
+        # start a TGI subprocess with the given parameter
+        args = f"text-generation-launcher --port 8080 --model-id {self.model} --huggingface-hub-cache /scratch {params}"
+        logger.info(f"Running TGI with parameters: {args}")
+        self.process = subprocess.Popen(args,
+                                        stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
+        for line in iter(self.process.stdout.readline, b""):
+            print(line.decode("utf-8"))
+            # wait for TGI to listen to the port
+            if b"Connected" in line:
+                break
+            if b"Error" in line:
+                raise Exception(f"Error starting TGI: {line}")
+
+        # continue to stream the logs in a thread
+        def stream_logs():
+            for line in iter(self.process.stdout.readline, b""):
+                print(line.decode("utf-8"))
+
+        if self.process.returncode is not None:
+            raise Exception("Error starting TGI")
+        self.thread = threading.Thread(target=stream_logs)
+        self.thread.start()
+
+    def stop(self):
+        logger.warning(f"Killing TGI with PID {self.process.pid}")
+        if self.process:
+            kill(self.process.pid)
+        if self.thread:
+            self.thread.join()
+
+
+class TGIDockerRunner(InferenceEngineRunner):
+    def __init__(self,
+                 model: str,
+                 image: str = "ghcr.io/huggingface/text-generation-inference:latest",
```

Comment:
are we sure that this runs on the correct docker image? Since it runs on `main` there shouldn't be any contention on concurrent builds 🤔

Reply:
You are right for PR runs that wouldn't work as we need to pinpoint a specific tag (that has to be done in `load_test.py`). For nightly, the current build process ensures that `latest` is a build run on `main`.